### ISSUE TITLE
containers: content control

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ArctosDB/arctos#565]
```
Don't allow the addition of non-position containers to containers that already 
contain positions, and vice-versa. (see: cryovials and positions in freezer box)
```

Original issue reported on code.google.com by `dust...@gmail.com` on 27 Jun 2012 at 2:40

**Comment by dustymc** [comment]
We need a fairly comprehensive list of what containers should allow/disallow before starting this - the code is very likely to be expensive and highly dependent so this is not a good place to let evolution run its course.

**Comment by dustymc** [comment]
Limit parent of and changes to institution and "trashcan" containers.

**Comment by dustymc** [comment]
Many (most?) errors seem to originate with users attempting to change the barcode of existing containers. I see no reason to ever change an existing barcode, and propose locking barcode from change.

**Comment by campmlc** [comment]
Barcodes occasionally need to be changed if a label becomes damaged or
torn. This can happen with the acrylic labels we use for freezer boxes and
cryovials.  Perhaps you could allow change but ask for verification, e.g. "
are you sure?"

**Comment by dustymc** [comment]
<img width="339" alt="screen shot 2015-10-01 at 1 33 02 pm" src="https://cloud.githubusercontent.com/assets/5720791/10233035/00bf1d2e-6841-11e5-9b77-52477be203d5.png">

exists to deal with damaged labels etc. - container replacement.

Suggestions/confirmations can be bypassed. The focus of this discussion should be rules - things that live on the tables and cannot be circumvented. (UI suggestions are always appreciated, but they won't stop the sorts of problem-causing conditions I see in the logs.)

**Comment by dustymc** [comment]
https://docs.google.com/spreadsheets/d/1v7kkKP2w0omkR_JNfDeCYU9PDHe8G32CvpUNxfHKv3M/edit?usp=sharing is current data. I think we could make serious progress by just listing allowable parents and contents. There's some tradeoff between allowing the things that actually happen - maybe someone dumped a moose in a freezer so we should allow collection_object-->freezer - and disallowing the things that probably didn't happen but got scanned that way anyway (there's probably no freezer-moose). 

Here's the start of a proposal:

- freezers contain only positions
- freezer racks contain only positions
- institutions contain only rooms
- legacy and unknown cannot do anything (if you're scanning it, you can figure out what it is)
- positions can only exist in (freezer, freezer rack, etc. - not "box") 
- identical parent/child container types (room-->room) are disallowed
- I'm not sure what pins can contain, but it probably does not include unit trays.
- slide boxes contain only slides
- vials cannot contain boxes
- herbarium folder contains only herbarium sheet
- herbarium sheet contains only collection object
- range case and case are used interchangeably - merge them
- tag contains only collection object

OK, I'll stop for now, there are lots more.

1) Does this seem like a useful direction?
2) If so, how do we get there?
3) Maybe we should discuss #745 first - whoever scanned a box into a cryovial may not be deterred by these rules.

**Comment by dustymc** [comment]
Disallow container movement by users who do not have access to (container.institution_acronym) joined through collection.institution_acronym

**Comment by campmlc** [comment]
good idea

**Comment by ccicero** [comment]
Most of those rules make sense, but I don't understand the one about positions not being allowed to exist in boxes. Our freezer boxes all have positions. ????

![box_positions](https://cloud.githubusercontent.com/assets/967275/10290360/7b74f11e-6b58-11e5-927c-8792e71ade78.PNG)

**Comment by dustymc** [comment]
UAM@ARCTOS> select container_type from ctcontainer_type order by container_type;

CONTAINER_TYPE
------------------------------------------------------------
Nalgene
bag
**box** <--- cardboard thing
cabinet
case
collection object
container label
cryovial
cryovial label
drawer
envelope
folder
freezer
**freezer box** <---- thing with positions
freezer rack
herbarium folder
herbarium sheet
institution
jar
legacy container
microplate
not recorded
pin
position
range
range case
room
shelf
slide
slide box
slide drawer
specimen label
tag
tray
unit tray
unknown
vial
vial unit tray
wood case

39 rows selected.

Elapsed: 00:00:00.01
UAM@ARCTOS>

**Comment by campmlc** [comment]
This does seem like a useful direction.

Re: box positions - I would very much like to request positions for boxes
which are not freezer boxes, because we use these for parasite vials stored
in freezers. In some cases,these are freezer boxes - but Arctos currently
does not accept that a vial can be scanned in a freezer box position. In
other cases, they are unit trays with larger vials, stored stacked in a -20
freezer. Currently I cannot scan any vials into a box or freezer box
position, and have approx 7,000 parasite vials locatable only by box
barcode.

**Comment by dustymc** [comment]
It's clear that proper care is NOT being taken with scans, and so we need to prevent vials being pinned and other such nonsense. There is no intention to prohibit any useful/intentional action. 

The "positions" app limits what it will accept - what happens there isn't a "rule," just a specialized form, and you can bypass those "restrictions" in any other container app. (The specialization lets the app make inferences and save you clicks - it'll magic "X label" into "X" container types, for example.)

As always, file an Issue if you need some functionality that doesn't exist.

The three boxes-with-positions are all KNWR - this probably needs to go to the larger group before we do anything very drastic.

**Comment by dustymc** [comment]
>no reason small things cannot be scanned into larger things, bypassing a level - but every reason why large things should not be scanned into small things.

If everything had a size we'd be done here. That has some issues in practice though - how big is a pin or tag? (Depends if you're pinning mosquitoes or tagging polar bear hides.)

Container_type is partially an attempt to add some organization when size fails.

>I have freezer boxes scanned directly into freezer racks (bypassing positions) for...various reasons....

Why do we have container type at all? My understanding is that it's to introduce predictability, and that can't work if there are exceptions. Without some hard rules, we WILL (already have) end up with a freezer rack of freezer racks. If limiting what we can put in certain types of containers isn't important, let's drop container_type altogether - it's not doing anything unless we USE it for something.

(I don't have much opinion on the details - rack-->box seems reasonable enough - but we do need to decide what it is that we're trying to do here.)

The "forced organization" is always easy to get around, and there may well be valid reasons to do so. Nothing prevents you from having a "shelf" that looks a lot like a freezer rack, and there are no restrictions on where you can put a shelf (eg, in a freezer) or what you can put in it (eg, boxes).

**Comment by dustymc** [comment]
Summary as of 2015-10-29:

See https://github.com/ArctosDB/DDL/blob/master/procedures/containerContentCheck.sql for current code

**Comment by campmlc** [comment]
I would like to suggest we start with the basic restrictions Dusty proposed and see what feedback is provided by users.

    freezers contain only positions

    [freezer racks contain only positions -defer this please until I can deal with a legacy issue;  it is useful to have direct rack to box relationships, so that in viewing the rack you immediately see the box contents rather than the position list. Also useful for smaller racks MLC]

    institutions contain only rooms
    legacy and unknown cannot do anything (if you're scanning it, you can figure out what it is)

    positions can only exist in (freezer, freezer rack, etc. - [please allow boxes to have positions MLC]

    identical parent/child container types (room-->room) are disallowed
    I'm not sure what pins can contain, but it probably does not include unit trays.

    slide boxes contain only slides [or slide positions MLC - although current slide position form translates to incorrect positions with most current browsers MLC, a very problematic issue]

    vials cannot contain boxes
    herbarium folder contains only herbarium sheet
    herbarium sheet contains only collection object
    range case and case are used interchangeably - merge them
    tag contains only collection object

**Comment by dustymc** [comment]
>[freezer racks contain only positions -defer this please until I can deal with a legacy issue;  it is useful to have direct rack to box relationships, so that in viewing the rack you immediately see the box contents rather than the position list. Also useful for smaller racks MLC]

All of the rules are applied on change; they won't affect legacy data.

We can "defer" anything, but the lack of control will probably continue to cause problems, eg:

UAM@ARCTOS> select distinct container_type from container where parent_container_id in (select container_id from container where container_type='freezer rack');

CONTAINER_TYPE
------------------------------------------------------------
position
freezer box
freezer rack

>[please allow boxes to have positions MLC]

box=random cardboard thing into which you can toss stuff.
{something} box=box with physical structure.

If you have a box with positions, you have something other than a "box." Maybe we need a new kind of container, or maybe we need new position-dimension values for freezer box. None of that is a reason to allow any of the 62,000 existing plain-boxes to contain positions.

>slide boxes contain only slides [or slide positions MLC 

As with "freezer box"es, we need to either enforce predictability or not, or maybe back way up and talk about container types. "Sometimes-except-when" "rules" are going to cause widespread confusion, and I'm incredibly hesitant to implement them. On the surface, I'd recommend limiting "slide boxes" to containing positions (how many positions do they contain? Maybe positions are not necessary at all??) and use plain ol' "box" if you're just tossing slides in.

More generally, I don't care how you map physical "container" descriptions to Arctos container types - boxes designed to hold slides don't have to be "slide boxes" in Arctos - but if you call it a "whatever" then it needs to follow the rules we've established for containers of type "whatever."

>- although current slide position form translates to incorrect positions with most current browsers MLC, a very problematic issue]

Hu? Screenshot/explanation?

**Comment by campmlc** [comment]
1) RE: boxes having positions - substitute "vial unit tray" and "unit tray" for box and it works for me

2) I've just now run into problems with the restriction on moving positions to a freezer that had previously had racks but no positions; I created the freezer positions independently and moved the racks to the new freezer positions in the universe, but now cannot move the positions + racks to the freezer. I know in the best of all possible worlds all container creation would be done in large batches in  a linear, logical, and orderly fashion according to some grand workflow or design, but the reality is that much of this has to be done piecemeal in an hour here or there or improvised to deal with legacy conversions. We either need a global training session demonstrating a master workflow, or we are going to have to ease restrictions and deal with the resulting errors. 

3) slide box positions are in two rows in the box with slide positions 1-50 on left, top to bottom, and 51-100 on right, top to bottom. This is how the boxes are labeled from the manufacturer. However, the slide box position form in Arctos reads left to right tabs instead of top to bottom returns. The result is that the position that is visible in the form (which correct) is different than the position indicated by clicking on part location in specimen results. (which is incorrect). My understanding is that it is a browser issue.

**Comment by dustymc** [comment]
>1) RE: boxes having positions 

Works for me - you might check in with Derek (who caused those container types to exist) before you get too deep, but there are currently no restrictions on them and I know of none being planned.

>2) moving positions

Yea, I think I finally understand the disconnect, will get you a new form ASAP.

>training

Kyndall and/or Aren are my recommendations. UAM has an incredibly solid object tracking system in place, and I never see container-errors from them which makes me think their non-database procedures are good too. No idea if the UAM system will completely work for other stuff, but at least the fundamentals should be comparable. 

We keep talking about "training modules"/demos/whatever, but it never happens. Is that something the AC could organize? Might even be worth it's own proposal??

> ease restrictions and deal with the resulting errors

I strongly recommend against this approach IF we want the sort of data that can only be produced with machine-readable labels (and lead to this issue in the first place). Part attribute "location" has no restrictions and is always available for "people-readable" data.

>3) slide positions

File a new Issue - this is a form/UI issue (and may possible/easier to address with technology that didn't exist when that form was built).

**Comment by dustymc** [comment]
Closing, file new Issue with specifics
